ignite/src/jepsen/ignite.clj

```diff
@@ -0,0 +1,161 @@
+
+(defn await-node-started
+    "Waits for the grid cluster started."
+    [node test]
+    (info node "Waiting for topology snapshot")
+    (let [tmp-file (File/createTempFile "jepsen-ignite-log" ".tmp")
+        tmp-file-path (.getCanonicalPath tmp-file)]
+    (try
+        ; Download server log file
+        (c/download logfile tmp-file-path)
+        (while (str/blank? (re-find (re-pattern (str "Topology snapshot \\[.*?, servers\\=" (count (:nodes test)) ","))
+            (slurp tmp-file-path))) (do
+                (Thread/sleep 1000)
+                (c/download logfile tmp-file-path)))
+    (finally
+        (.delete tmp-file)))))
+
+(defn nuke!
+    "Shuts down server and destroys all data."
+    [node test]
+    (c/su
+        (meh (c/exec :pkill :-9 :-f "org.apache.ignite.startup.cmdline.CommandLineStartup"))
+        (c/exec :rm :-rf server-dir))
+    (info node "Apache Ignite nuked"))
+
+(defn configure [addresses client-mode]
+    "Creates a config file."
+    (let [tmp-file (File/createTempFile "jepsen-ignite-config" ".xml")
+          tmp-file-path (.getCanonicalPath tmp-file)]
+        (spit tmp-file-path (clojure.string/replace (slurp "resources/apache-ignite.xml.tmpl")
+            #"##addresses##|##instancename##|##clientmode##"
+            { "##addresses##" (clojure.string/join "\n" (map #(str "\t<value>" % ":47500..47509</value>") addresses))
+            "##instancename##" (.getName tmp-file)
+            "##clientmode##" (str client-mode) }))
+        tmp-file))
+
+(defn configure-server [addresses node]
+    "Creates a server config file and uploads it to the given node."
+    (let [src-file (configure addresses false)]
+        (c/upload src-file (str server-dir "server-ignite-" node ".xml"))
+        (.delete src-file)))
+
+(defn configure-client [addresses]
+    "Creates a client config file."
+    (configure addresses true))
+
+(defn db
+    "Apache Ignite cluster life-cycle."
+    [version]
+    (reify db/DB
+        (setup! [_ test node]
+            (info node "Installing Apache Ignite" version)
+            (debian/install-jdk8!)
+            (cu/ensure-user! user)
+            (let [url (str "https://archive.apache.org/dist/ignite/" version "/apache-ignite-" version "-bin.zip")]
+                (cu/install-archive! url server-dir))
+            (configure-server (:nodes test) node)
+            (c/exec :chown :-R (str user ":" user) server-dir)
+            (start! node test)
+            (await-node-started node test)
+            (jepsen/synchronize test))
+
+        (teardown! [_ test node]
+            (nuke! node test))
+
+        db/LogFiles
+            (log-files [_ test node]
+            [logfile])))
+
+(defn r   [_ _] {:type :invoke, :f :read, :value nil})
+(defn w   [_ _] {:type :invoke, :f :write, :value (rand-int 1000)})
+
+(defrecord Client [conn cache config]
+    client/Client
+    (open! [this test node]
+        (let [config (configure-client (:nodes test))
```

Comment:
You generally want to avoid letting a client talk to multiple nodes; this is a good way to miss concurrency violations.